compiler/optimizer/LocalOpts.cpp

```diff
@@ -3661,7 +3661,7 @@ int32_t TR_EliminateRedundantGotos::process(TR::TreeTop *startTree, TR::TreeTop
       // Place an asynccheck as the first treetop of the successor, if there was one in the removed block
       // This is necessary as placing it in the predecessor may result in a seperation from its OSR guard
       //
-      if (asyncMessagesFlag && comp()->isOSRTransitionTarget(TR::postExecutionOSR))
+      if (asyncMessagesFlag && comp()->getHCRMode() == TR::osr)
```

Comment:
This is the case where the asynccheck should be inserted under NextGenHCR, as it won't invalidate an assumption without an OSR guard. All cases being treated the same was a bug.